### DISCUSSION TITLE
SubscribeForNotifications() gives out-of-bounds changes indexes

### DISCUSSION THREAD

**Discussion by jrlaff** [realm/realm-dotnet#2736] [Q&A]
Hi,

Using Realm's `CollectionExtensions.SubscribeForNotifications()` to watch for changes in a query results collection, I often receive a change notification with a `changes` ChangeSet that contains indexes that are out of the bounds of the collection. It happens quite regularly, but I haven't been able to pinpoint a precise scenario that reproduces the issue.

More precisely, `changes` contains, for example, 5 `NewModifiedIndices` (0, 1, 2, 3, 4) but `sender` only contains 4 items. It does not contain any other index (`InsertedIndices` and `DeletedIndices` are empty).

No need to say that accessing the sender using those indices triggers an exception.

How am I supposed to handle this?

Any help appreciated. Thanks.

**Comment by papafe** [comment]
Hi @jrlaff, this sounds like a bug. I understand that you didn't manage to find precisely when it happens, but if you manage to do so it will be really helpful for us to understand what is going on. Even if it does not happen all the time, having a project with which we can reproduce the error will definitely help.

**Reply by jrlaff** [reply]
There may be something. I notice that the issue occurs when I open a transaction inside a changes notification. Is it something that could cause trouble? I basically have (sub methods calls simplified):

```
private void OnResultsChanged(IRealmCollection<MyObject> sender, ChangeSet changes, Exception error)
{
    [...]
    using (var transaction = realm.BeginWrite())
    {
        [...]
        transaction.Commit();
    }
}
```

Is it maybe forbidden to alter the Realm's contents within a changes notification callback?

**Reply by papafe** [reply]
Do you get those issues with non existing indexes in the same notification blocks where you create transactions? Would it be possible for you to show when/how you access the indexes?

Also, can I ask you why you are opening a transaction inside the notification block? It would be interesting to know what is your use case and what you are trying to accomplish.

**Reply by jrlaff** [reply]
I need to be able to react to changes in some tables to build objects based on these changes.

A typical use-case is a telephony application that would handle Call objects that come and go as activity requires, and needs to maintain a Call History feature that requires detecting the creation/update/removal of a call to create a Call log object that saves what happened. In this regards, a module in the application would need to listen to a Call object creation and later update/removal (call disappears without being answered) to create a new Call log object (missed call).

Basic applications that present objects and allow creation/update/delete operations do not probably require such mechanisms, but IMO any serious application that needs to provide more complex operations need this.

In my example, I notice that the callback triggers a recursive call to the same callback from the transaction and issues arise as the changes parameter cannot be used against the collection because the recursive call changes the collection. I haven't been able to determine why most of the objects seem to be included in the changes and specially the same objects as the ones used in the original collection changes subscription, even if the transaction does not directly update the same objects. I feel that objects are marked as updated even if none of their direct properties are modified, but nested objects are...

**Reply by nirinchev** [reply]
Do you get the out of bound indexes before or after you open the transaction? I.e. if you check the size of the collection before you call `realm.Write`, is it less than the larger index in the collection?

Generally speaking, it's not recommended to open transactions inside a notification callback because that may result in unexpected behavior. To try and explain the issue, every time a Realm file changes, an internal version is created and notifications are emitted by calculating the changes between the previous and the current version. Calculation of changes is done in the background though, so it's possible to have more updates that occur while we're delivering notifications. The notification delivery mechanism ensures that within the notification callback you're seeing the version that we've calculated notifications for. Let's walk this through step by step:

1. Your realm is at version `1`.
2. You make some changes bringing it to version `2`.
3. Notifications are calculated for the delta `1` -> `2` and the notification callback is emitted.
4. Even if in the background, a write happened bringing the Realm to version `3`, the `sender` in the callback is pinned at version `2` to make sure you won't see nonsensical data, such as getting modification notifications for deleted objects.

Now, if you do a `realm.Write` inside the notification callback, you're forcing the collection to advance to the latest version as you can't write to an outdated version. You're also forcing notifications to be delivered for the changes between `2` and `3`. So what happens when you do `realm.Write` in the callback is:

1. Another callback will be invoked before `realm.Write` returns with the changes from `2` to `3`, which may do its own `Write` and trigger yet another callback if there were background changes.
2. When you go back to the original callback, `sender` will now be at version `3`, but the `changes` parameter will describe the changes in version `2`. This means that if the collection was further modified in version `3`, you're bound to see unexpected results, such as indexes being out of range.

If you absolutely must use write transactions inside a callback, the recommendation would be to preprocess all the values that you need before opening a write transaction. It's hard to give you an example without knowing your use case or seeing your code, but something like:

```csharp
query.SubscribeForNotifications((sender, changes, error) =>
{
    // bad, don't do this
    foreach (var index in changes.ModifiedIndices)
    {
        var obj = sender[index];
        realm.Write(() =>
        {
            obj.Foo = "bar";
        });
    }

    // Instead do this
    // Get all objects ahead of time to avoid advancing the collection to the next version
    var modifiedObjs = changes.ModifiedIndices.Select(i => sender[i]).ToArray();

    // Do only one write at the end of the notification block
    realm.Write(() =>
    {
        foreach (var obj in modifiedObjs)
        {
            // Check the object hasn't been deleted when we advanced the version
            if (!obj.IsValid)
            {
                continue;
            }

            // Don't modify an object if it already has the correct value. Otherwise
            // you risk creating an endless feedback loop.
            if (obj.Foo != "bar")
            {
                obj.Foo = "bar";
            }
        }
    });
});
```

**Reply by jrlaff** [reply]
Thanks @nirinchev, your answer definitely helps getting a sharper picture of how it works.

I indeed changed my callback body to preprocess the changes and get the objects in an array. This got me **rid of the exceptions** I had. I also opened one **single transaction** to process the sub-changes I have to do, after reading the guidelines that encourage to open a **single transaction** around a loop instead of multiple smaller transactions.

One more question regarding my need to **store previous versions of objects** so that I can compare what has changed in each object and find out which object got deleted: **is there any standard way to get a copy of a realm object that is detached, and thus can be safely stored for later use (that use being comparing to the next version to see what has changed)?** I have so far implemented dummy field-by-field copy methods, but that looks so wrong and I don't really know what to do with nested objects.

Thank you very much.

**Reply by nirinchev** [reply]
Generally, copying things field-by-field is the best way to go about it as it gives you the most flexibility in terms of what you need to copy. There are some amazing 3rd party libraries, such as [Automapper](http://automapper.org) that automate a lot the boilerplate code there. For example, if you know that you only care about 2-3 fields, you can configure automapper to produce detached objects with only these fields populated.

Alternatively, you can freeze an object using the [`.Freeze()`](https://docs.mongodb.com/realm-sdks/dotnet/latest/reference/Realms.FrozenObjectsExtensions.html#Realms_FrozenObjectsExtensions_Freeze__1___0_) extension method. It will keep a reference to an object at a particular version and you can read it even after the original object has been deleted. One caveat here is that you have to be careful not to keep those references forever, otherwise the filesize of the Realm may grow significantly. This is caused by the fact Realm will currently keep all intermediate versions between the oldest pinned (frozen) version and the current one. So if you pinned an object at version 1, then did 1000 transactions alternatingly changing a boolean property, Realm would need to keep 1000 versions around, even though the end state is the same as the frozen state. This is something we're looking to optimize, but don't have a timeframe for it yet.

**Reply by jrlaff** [reply]
Thank you @nirinchev. I am looking at Automapper and it seems very handy.

I would have marked your messages as answer but it can only be done on direct replies to the main thread...

**Reply by nirinchev** [reply]
No worries, the important part is that we solve your problem 😄
